TripalCultivate/TripalCultivate
# G1.25 experiment content type updates

**Issue #25**

**Previous PR #26** (Closed due to the scope of changes and length of original)

## Motivation
We have determined that our original approach of having multiple Experiment Type content types (i.e. Field Experiment, Greenhouse Experiment, Biochemical Assay) does not work well since many experiments may include components of two experiment types. For example, the microbiome work involves sampling from a field experiment and then sequencing to determine the species composition. Another example of this would be a food quality experiment where we need to describe the biochem assay and the way the seeds were obtained.

## What does this PR do?

As such, this PR brings together all the fields from the original Experiment Type content types into a single content type. It also strives to organize these fields in a less overwhelming way and only have one copy of any common fields (i.e. name, objectives, summary).

More specifically, this PR

- [x] Removes the individual experiment type content types from `tripal.tripalentitytype_collection.trpcultivate_experiments.yml` in favour of a single `research_experiment` content type
- [x] Rearranges, names and combines any common fields across the 4 content types into a single content type in `tripal.tripalfield_collection.trpcultivate_experiments.yml` making sure not to loose any experiment type specific instructions.
- [x] Adds a default layout for the display (page view) of an experiment in `tripal_layout.tripal_layout_default_view.trpcultexperiment_layout.yml` to automate configuring the order of the many fields.
- [x] Adds a default layout for the form (edit) of an experiment in `tripal_layout.tripal_layout_default_form.trpcultexperiment_layout.yml` to automate configuring the order of the many fields.
- [x] Works to provide a consistent and intuitive organization of fields for both the experiment form and display.
- [x] Changes the "Germplasm Species" field to a "Germplasm Scientific Name" field to handle the situation where there are multiple genus'.
- [x] Add support for automatically creating markup and Entity Reference fields which cannot be added via the Tripal Fields Collection since they are not Tripal Fields.
- [x] Organize into vertical tabs to make the form less overwhelming
- [x] Change the organization to map better to how an experiment is designed with each tab representing steps of the process.
- [x] Add better group instructions for various panes.
- [x] Auto-configure some of the field widget size + options for a better experience (e.g. configure property to be a select list of options)
- [x] Auto-apply layouts.
- [x] Fix automated tests.

### What to Expect

Expand the following section to see screenshots of the form in order to see what to expect.

<details>
<summary>What to Expect</summary>

  ![Objectives](https://github.com/user-attachments/assets/98d36a4a-041c-4742-bf1f-6e1ef9089c07)
  ![Attribution](https://github.com/user-attachments/assets/e90f22d7-9c83-41d6-8c7c-0116f41f1966)
  ![Design](https://github.com/user-attachments/assets/0f9dcb9f-371a-4159-8d99-6be9dc82186c)
  ![Methodology](https://github.com/user-attachments/assets/bc7d4150-41f4-4abb-bfda-34be03e083f9)
  ![References](https://github.com/user-attachments/assets/57d3c260-c70e-4119-b47d-1f8e335216a5)

</details>

## Testing

### Automated Testing




### Manual Testing


1. Create a fresh container on this branch. If you are using VSCode devcontainers then just click "Reopen in container" and if not, see the instructions at the bottom.
2. Create a number of contacts by going to Top Bar > Tripal > Content > Add Tripal Content > Contact. We will use these as the researchers and organization in step 4.
3. Create a Grant and a study using the same approach.
4. Open the Create Research Experiment form by going to Top Bar > Tripal > Content > Add Tripal Content > Research Experiment
5. Check that the form layout is intuitive, fill it out and make sure it works this time 🙈 

#### Build container from scratch

```
cd ~/Dockers
git clone https://github.com/TripalCultivate/TripalCultivate trpcultBase-35
cd trpcultBase-35
git checkout g1.25-experimentContentTypeUpdates
docker build --tag=trpcultivate-base:review35 ./
docker run --publish=80:80 -tid --name=base35 --volume=`pwd`:/var/www/drupal/web/modules/contrib/TripalCultivate trpcultivate-base:review35
```
